python/phases.py

```diff
@@ -0,0 +1,107 @@
+from typing import Union, Set, Tuple
+import pathlib
+import os
+import copy
+import sys
+import time
+
+import lxml.html
+
+from traverse import traverse_and_replace
+
+PYTHON_PATH = pathlib.Path(__file__).parent.parent.resolve()
+CORE_ELEMENTS_PATH = (PYTHON_PATH / "elements").resolve()
+SAVED_PATH = copy.copy(sys.path)
+
+
+# def resolve_element(element_name: str, context: dict) -> dict:
+#     if element_name in context["course_elements"]:
+#         return context["course_elements"][element_name]
+#     elif element_name in context["c_elements"]:
+
+
+def render(data: dict, context: dict) -> Tuple[str, Set[str]]:
+    # This will be a string consisting of `question.html` with Mustache templating applied.
+    html = context["html"]
+
+    # This will be a dict mapping an element name to information about them.
+    elements = context["elements"]
+
+    # This will be a dict mapping an element name to a dict of extensions for that element.
+    element_extensions = context["element_extensions"]
+
+    # This will track which elements have been rendered.
+    rendered_elements: Set[str] = set()
+
+    total_time = 0
+
+    def render_element(element: lxml.html.HtmlElement) -> Union[str, None]:
+        if element.tag not in elements:
+            return None
+
+        start = time.time()
```

Comment:
Saw this somewhere else too, I'm assuming that this is to time the renderer to get performance data? Should use time.perf_counter to get more accurate results.